### PR TITLE
add opengraph tags

### PR DESCRIPTION
**Pull request by florpor**
using a unified serve script which serves both static react-built assets and the api

### CODE REVIEW

File: bin/serve [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+const apiService = require('../src/services/api');
+const config = require('../src/config.json');
+
+
+// create static app
+const urlencoded = BodyParser.urlencoded({
+  extended: false,
+});
+const staticApp = Express();
```
Comment (dortheimer) [inline bin/serve]: Let's not have two apps on two ports.
Better that we have one that serves both.

File: bin/serve [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        pageDescription = `תב"ע ב${plan.data.data.PLAN_COUNTY_NAME}`;
+
+      res.render('index', {
+        layout: false, 
+        pageTitle: `${plan.data.PL_NUMBER} - ${plan.data.PL_NAME} - מעירים`, 
+        pageDescription: pageDescription,
+        pageImage: config.opengraph.basePageImageURL,
+      });
+    }).catch(error => {
+      console.error('failed to fetch plan details for opengraph');
```
Comment (dortheimer) [inline bin/serve]: Use logger. Console is evil.

File: bin/serve [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      else
+        pageDescription = `תב"ע ב${plan.data.data.PLAN_COUNTY_NAME}`;
+
+      res.render('index', {
+        layout: false, 
+        pageTitle: `${plan.data.PL_NUMBER} - ${plan.data.PL_NAME} - מעירים`, 
+        pageDescription: pageDescription,
+        pageImage: config.opengraph.basePageImageURL,
+      });
+    }).catch(error => {
```
Comment (dortheimer) [inline bin/serve]: Use an error handler using next()

File: bin/serve [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+staticApp.set('views', staticRootDir);
+staticApp.engine('html', ExpressHandlebars({extname: '.html', layoutsDir: staticRootDir}));
+staticApp.set('view engine', 'html');
+
+staticApp.get('/plan/:planId/:planName?', function(req, res) {
+  const { planId, planName } = req.params;
+
+  // fetch plan details to populate og tags
+  apiService.get('/plan/' + planId)
+    .then(plan => {
```
Comment (dortheimer) [inline bin/serve]: put the routes in a routes file

File: bin/serve [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        layout: false, 
+        pageTitle: 'תוכנית בניה - מעירים', 
+        pageDescription: '',
+        pageImage: config.opengraph.basePageImageURL,
+      });
+    });
+});
+
+// all non-plan pages have static og tag values
+staticApp.get('*', function(req, res) {
```
Comment (dortheimer) [inline bin/serve]: This doesn't make sense since we should return errors for unexisting urls.

File: src/config.json [hunk 6]
```diff
@@ -2,5 +2,8 @@
   "axios": {
     "baseURL": "http://localhost:3001",
     "withCredentials": true
+  },
+  "opengraph": {
+    "basePageImageURL": "http://localhost:3000/favicon.ico"
```
Comment (dortheimer) [inline src/config.json]: Use the baseUrl configuration and add favicon hardcoded.

File: public/index.html [hunk 7]
```diff
@@ -13,6 +13,12 @@
         <meta name="description" content="שירות המעדכן אתכם בכל פעם שמתפרסמת תוכנית בניין עיר חדשה באיזורכם" />
         <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />
 
+        <meta property="og:locale" content="he_IL" />
```
Comment (dortheimer) [inline public/index.html]: Locale should be in the configuration.

File: public/index.html [hunk 8]
```diff
@@ -13,6 +13,12 @@
         <meta name="description" content="שירות המעדכן אתכם בכל פעם שמתפרסמת תוכנית בניין עיר חדשה באיזורכם" />
```
Comment (dortheimer) [inline public/index.html]: The description and title should be also dynamic

File: package.json [hunk 9]
```diff
@@ -93,6 +94,7 @@
   "scripts": {
     "start": "BROWSER=none ./node_modules/react-scripts/bin/react-scripts.js start",
     "server": "node bin/api",
-    "build": "./node_modules/react-scripts/bin/react-scripts.js build"
+    "build": "./node_modules/react-scripts/bin/react-scripts.js build",
+    "serve": "node bin/serve"
```
Comment (dortheimer) [inline package.json]: Should be removed.

File: bin/serve (hunk 2 above)
Comment (florpor) [reply bin/serve]: done

File: bin/serve (hunk 3 above)
Comment (florpor) [reply bin/serve]: we shouldn't use next() because we want to show the react 404 page

File: bin/serve (hunk 4 above)
Comment (florpor) [reply bin/serve]: done

File: bin/serve (hunk 5 above)
Comment (florpor) [reply bin/serve]: same as above - we shouldn't return errors for the static app so the react app loads and shows the react 404 page

File: src/config.json (hunk 6 above)
Comment (florpor) [reply src/config.json]: removed

File: bin/serve (hunk 1 above)
Comment (florpor) [reply bin/serve]: they are loaded into one main app

File: public/index.html (hunk 7 above)
Comment (florpor) [reply public/index.html]: done

File: api/staticRoutes.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        layout: false, 
+        pageTitle: 'תוכנית בניה - מעירים', 
+        pageDescription: '',
+        pageImage: `${req.protocol}://${req.get('host')}/favicon.ico`,
+      });
+    });
+});
+  
+// all non-plan pages have static opengraph tag values
+Router.get('*', function(req, res, next) {
```
Comment (dortheimer) [inline api/staticRoutes.js]: We have some specific pages, let's define them so we can respond with 404 errors when links are broken. This is very bad for seo....

File: api/staticRoutes.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    pageDescription: 'רוצים לדעת אם הגינה הציבורית שלכם עומדת להפוך למגדל?',
+    pageImage: `${req.protocol}://${req.get('host')}/favicon.ico`,
+  });
+});
+
+/**
+ * Convert date object to string with "dd/MM/yyyy" format.
+ */
+
+function dateStrFormat(dateStr) {
```
Comment (dortheimer) [inline api/staticRoutes.js]: Better us arrow functions.

File: api/staticRoutes.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  try {
+    const dt = new Date(dateStr);
```
Comment (dortheimer) [inline api/staticRoutes.js]: You should use moment for this.

File: api/staticRoutes.js [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+  const { planId } = req.params;
+
+  // fetch plan details to populate opengraph tags
+  Plan.fetchByPlanID(planId)
+    .then(plan => {
+      let pageDescription;
+
+      // some plans don't have depositing dates
+      if (plan.attributes.data.DEPOSITING_DATE)
+        pageDescription = `תב"ע ב${plan.attributes.data.PLAN_COUNTY_NAME} אשר הופקדה ב ${dateStrFormat(plan.attributes.data.DEPOSITING_DATE)}`;
```
Comment (dortheimer) [inline api/staticRoutes.js]: The strings should be in the template.
You can have an if condition with mustache.

File: api/staticRoutes.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        pageDescription: pageDescription,
+        pageImage: `${req.protocol}://${req.get('host')}/favicon.ico`,
+      });
+    }).catch(error => {
+      Log.error('failed to fetch plan details for opengraph');
+      Log.error(error);
+
+      res.render('index', {
+        layout: false, 
+        pageTitle: 'תוכנית בניה - מעירים',
```
Comment (dortheimer) [inline api/staticRoutes.js]: Localized strings must be in the template. Just make an error template.

File: api/staticRoutes.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
+const Router = require('express').Router();
+
+const Log = require('./lib/log');
+const Plan = require('./model/plan');
+const config = require('../src/config.json');
+
+// all plan pages should render opengraph tags
+Router.get(['/plan/:planId', '/plan/:planId/*'], function(req, res, next) {
```
Comment (dortheimer) [inline api/staticRoutes.js]: Better use an arrow function.

**Pull request merged**